Add SPI for setting sites to be excluded from website data deletion

szewai opened the pull request:
#### 63ef7c135a1751344522a4662e130e9bd76191ef
<pre>
Add SPI for setting sites to be excluded from website data deletion
<a href="https://bugs.webkit.org/show_bug.cgi?id=276786">https://bugs.webkit.org/show_bug.cgi?id=276786</a>
<a href="https://rdar.apple.com/130761453">rdar://130761453</a>

Reviewed by Chris Dumez.

With existing implementation, website data is removed by registrable domain, so UI process sends persisted domains
instead of sites to network process.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::domainsExemptFromWebsiteDataDeletion const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setPersistedDomains):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setPersistedDomains):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::NetworkSession):
(WebKit::NetworkSession::forwardResourceLoadStatisticsSettings):
(WebKit::NetworkSession::setPersistedDomains):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.h:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _persistedSites]):
(-[WKWebsiteDataStore _setPersistedSites:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setPersistedSiteURLs):
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::persistedSiteURLs const):

Canonical link: <a href="https://commits.webkit.org/281200@main">https://commits.webkit.org/281200@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f030004db22b86cae5a4f33816d5a98b9b2e5c98

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59102 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38430 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11599 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62881 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9545 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46081 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9752 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47934 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6823 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61132 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35947 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51149 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/28785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32670 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8549 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54625 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8695 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64583 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3014 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8655 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55260 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51151 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55373 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13080 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2532 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/34263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36428 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/Site.h:
```diff
@@ -38,6 +38,10 @@ class Site {
     Site(const Site&) = default;
     Site& operator=(const Site&) = default;
 
+    Site(String&&, WebCore::RegistrableDomain&&);
```
We should not omit the first parameter name, it is not obvious it is a protocol.

szewai replied:
This is removed.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -793,6 +793,11 @@ HashSet<RegistrableDomain> ResourceLoadStatisticsStore::domainsExemptFromWebsite
 
+    for (auto& site : m_persistedSites) {
+        if (!site.isEmpty())
+            result.add(site.domain());
```
So we only need registrable domains? If so, why are we sending Site objects to the NetworkProcess instead of RegistrableDomains?

szewai replied:
Yes, I made UI process only sends domains to network process.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -660,6 +661,28 @@ - (void)_setStorageSiteValidationEnabled:(BOOL)enabled
+- (NSArray<NSURL *> *)_persistedSites
+{
+    auto sites = _websiteDataStore->persistedSites();
+    auto result = adoptNS([[NSMutableArray alloc] initWithCapacity:sites.size()]);
```
RetainPtr

szewai replied:
Updated.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -472,6 +472,9 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
     bool operator==(const WebsiteDataStore& other) const { return (m_sessionID == other.sessionID()); }
     void resolveDirectoriesAsynchronously();
 
+    HashSet<Site> persistedSites() const { return m_persistedSites; }
```
should return a `const HashSet<Site>&`

szewai replied:
Updated.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1829,6 +1829,13 @@ void WebsiteDataStore::setStorageSiteValidationEnabled(bool enabled)
+void WebsiteDataStore::setPersistedSites(HashSet<Site>&& persistedSites)
+{
+    m_persistedSites = WTFMove(persistedSites);
+    if (RefPtr networkProcess = networkProcessIfExists())
```
In the case where the network process isn't launched yet, we need to pass the persisted domains to the network process during launch (via the `WebsiteDataStoreParameters` probably). This is also important in case the network process crashes and relaunches.

szewai replied:
Nice catch; added to `ResourceLoadStatisticsParameters`.

cdumez commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -81,6 +81,7 @@ class NetworkResourceLoader;
 class NetworkSocketChannel;
 class NetworkStorageManager;
 class ServiceWorkerFetchTask;
+class Site;
```
nit: No longer necessary.

cdumez commented on Source/WebKit/Shared/Site.h:
```diff
@@ -57,6 +57,7 @@ class Site {
         static bool equal(const Site& a, const Site& b) { return a == b; }
         static const bool safeToCompareToEmptyOrDeleted = false;
     };
+    String string() const;
```
nit: I don't think we need this anymore.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -34,6 +34,7 @@
 #import "NetworkProcessProxy.h"
 #import "RestrictedOpenerType.h"
 #import "ShouldGrandfatherStatistics.h"
+#import "Site.h"
```
I don't think we need this anymore.

szewai committed:
Add SPI for setting sites to be excluded from website data deletion
https://bugs.webkit.org/show_bug.cgi?id=276786
rdar://130761453

Reviewed by Chris Dumez.

With existing implementation, website data is removed by registrable domain, so UI process sends persisted domains
instead of sites to network process.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::domainsExemptFromWebsiteDataDeletion const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setPersistedDomains):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setPersistedDomains):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::NetworkSession):
(WebKit::NetworkSession::forwardResourceLoadStatisticsSettings):
(WebKit::NetworkSession::setPersistedDomains):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.h:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _persistedSites]):
(-[WKWebsiteDataStore _setPersistedSites:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setPersistedSiteURLs):
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::persistedSiteURLs const):

Canonical link: https://commits.webkit.org/281200@main
